stablehlo/dialect/Base.cpp

```diff
@@ -176,5 +175,137 @@ Attribute boundsToEncoding(Attribute prototype, ArrayRef<int64_t> bounds) {
+  return std::make_pair(ShapedType::kDynamicSize, ShapedType::kDynamicSize);
+}
+
+LogicalResult inferMostSpecificType(Optional<Location> location,
+                                    ValueTypeRange<ValueRange> inputTypes,
+                                    SmallVectorImpl<Type>& inferredReturnTypes,
+                                    int64_t concatenateDim) {
+  SmallVector<RankedTensorType> rankedTypes;
+  for (auto inputType : inputTypes)
+    if (auto rankedType = inputType.dyn_cast<RankedTensorType>())
+      rankedTypes.push_back(rankedType);
+  if (rankedTypes.empty()) {
+    inferredReturnTypes.push_back(inputTypes[0]);
+    return success();
+  }
+
+  auto rank = rankedTypes[0].getRank();
+  // BoundedDialectInterface* dialect = nullptr;
+  SmallVector<int64_t> inferredDimSizes(rank, ShapedType::kDynamicSize);
+  SmallVector<int64_t> inferredBounds(rank, ShapedType::kDynamicSize);
+
+  bool anyInputHaveBounds = false;
+  for (const auto& it : llvm::enumerate(rankedTypes)) {
+    RankedTensorType rankedType = it.value();
+    ArrayRef<int64_t> bounds = encodingToBounds(rankedType.getEncoding());
+    if (!bounds.empty()) anyInputHaveBounds = true;
+
+    // Init the inferredDimSizes[dim] / inferredBounds[dim] with rankedTypes[0]
+    if (it.index() == 0) {
+      for (int dim = 0; dim < rank; ++dim) {
+        inferredDimSizes[dim] = rankedType.getDimSize(dim);
+        inferredBounds[dim] =
+            bounds.empty() ? ShapedType::kDynamicSize : bounds[dim];
+      }
+      continue;
+    }
+
+    // Infer each dim for current rankedType (from inputTypes[1:end])
+    for (int dim = 0; dim < rank; ++dim) {
+      std::pair<int64_t, int64_t> inferredDimAndBound;
+
+      int64_t leftSize = inferredDimSizes[dim];
+      int64_t rightSize = rankedType.getShape()[dim];
+      int64_t leftBound = inferredBounds[dim];
+      int64_t rightBound =
+          bounds.empty() ? ShapedType::kDynamicSize : bounds[dim];
+      if (dim == concatenateDim) {
+        inferredDimAndBound = inferConcatenatedDimAndBound(
+            leftSize, rightSize, leftBound, rightBound);
+      } else {
+        auto inferredDimAndBoundOrErr = inferMergedDimAndBound(
+            location, dim, leftSize, rightSize, leftBound, rightBound);
+        if (failed(inferredDimAndBoundOrErr)) return failure();
+        inferredDimAndBound = *inferredDimAndBoundOrErr;
+      }
+      inferredDimSizes[dim] = inferredDimAndBound.first;
+      inferredBounds[dim] = inferredDimAndBound.second;
+    }
+  }
+
+  inferredReturnTypes.push_back(RankedTensorType::get(
+      inferredDimSizes, rankedTypes[0].getElementType(),
+      boundsToEncoding(
+          rankedTypes[0].getEncoding(),
+          // Empty array as argument is an indicator to boundsToEncoding() that
+          // there are no bounds at all in inputs, thus sparsity attributes will
+          // be included in the return type
+          anyInputHaveBounds ? inferredBounds : llvm::ArrayRef<int64_t>({}))));
```

Comment:
Seems like we didn't propagate sparse encoding for binary ops earlier? What should be the logic there in case of different sparse encoding for lhs and rhs? Could we add a test there? Same for concatenate op.

Reply:
We do, and used the rankTypes[0].encoding() for the return encoding. 
Now for both in Binary and Concat op, we both keep using `rankTypes[0].encoding()`. But, I think this is an open question: let's track in https://github.com/openxla/stablehlo/issues/407